### PR TITLE
Handle the `MP_TUPLE` extension in the buffer source's `next` method

### PR DESCRIPTION
**Pull request by CuriousGeorgiy**
In scope of tarantool/tarantool#8147, a new context-dependent extension for box tuples, `MP_TUPLE`, is introduced. The buffer source uses a buffer with raw MsgPack, which does not allow for passing the context required for decoding `MP_TUPLE`, so, in order to decode it, we need to manually skip the extension header and the tuple format identifier to get to the tuple data and create a tuple. We can ignore the tuple format identifier (and the tuple format that was originally sent for this tuple), since the format is provided by the merger itself.

### CODE REVIEW

File: src/merger/merger.c [hunk 1]
```diff
@@ -597,6 +597,19 @@ luaL_merge_source_buffer_next(struct merge_source *base,
...
+	 */
+	if (mp_typeof(*tuple_beg) == MP_EXT) {
+		uint32_t len;
+		int8_t type;
+		mp_decode_ext(&tuple_beg, &type, &len);
+		assert(tuple_beg + len == tuple_end);
+		assert(mp_typeof(*tuple_beg) == MP_UINT);
+		/* Skip the tuple format identifier. */
+		mp_decode_uint(&tuple_beg);
+	}
```
Comment (Totktonada) [inline src/merger/merger.c:612]: Shouldn't we perform this logic only if the the extension is of type MP_TUPLE?

I know that it wouldn't change the behavior much due to https://github.com/tarantool/tarantool/issues/5048. However, it at least wouldn't make the problem deeper.

Comment (CuriousGeorgiy) [reply src/merger/merger.c:612]: How do we know we know what is the code of `MP_TUPLE`? Should I hardcode it?

Comment (CuriousGeorgiy) [reply src/merger/merger.c:612]: > Shouldn't we perform this logic only if the the extension is of type MP_TUPLE?

I expected that only the `MP_TUPLE` extension can lead to this branch (in the embedded version I have an assertion on this). Not clear what to do if we get a different extension type.

Comment (Totktonada) [reply src/merger/merger.c:612]: > How do we know we know what is the code of `MP_TUPLE`? Should I hardcode it?

Sure. The extension codes are part of the public binary protocol. See https://www.tarantool.io/en/doc/latest/dev_guide/internals/msgpack_extensions/

> I expected that only the `MP_TUPLE` extension can lead to this branch (in the embedded version I have an assertion on this). Not clear what to do if we get a different extension type.

That's so if the input buffer is well-formed. However, we shouldn't crash if it is ill-formed. We should report an error instead.

Comment (Totktonada) [reply src/merger/merger.c]: > I added an `mp_extension_type` enum for the `MP_TUPLE` code, and now I set a diagnostic and return an error when the extension type is wrong. 

It looks correct for me. Thanks!

**Review by Totktonada — APPROVED** [review]
Looks OK to me, but I would fix CI before merge and release it (it seems, it is on me).

**Comment by Totktonada** [comment]
Rebased after PR #38.

**Pull request merged**